app_check/src/ios/app_check_ios.mm

```diff
@@ -91,13 +92,60 @@ - (id)initWithProviderFactory:(firebase::app_check::AppCheckProviderFactory* _No
 
 @end
 
+@implementation AppCheckNotificationCenterWrapper
+
+// A collection of NSValues containing AppCheckListener*
+NSMutableArray* listeners_;
+
+- (id)init {
+  self = [super init];
+  if (self) {
+    listeners_ = [[NSMutableArray alloc] init];
+    [[NSNotificationCenter defaultCenter] addObserver:self
+                                             selector:@selector(appCheckTokenDidChangeNotification:)
+                                                 name:FIRAppCheckAppCheckTokenDidChangeNotification
+                                               object:nil];
+  }
+  return self;
+}
+
+- (void)addListener:(firebase::app_check::AppCheckListener* _Nonnull)listener {
+  [listeners_ addObject:[NSValue valueWithPointer:listener]];
+}
+
+- (void)removeListener:(firebase::app_check::AppCheckListener* _Nonnull)listener {
+  [listeners_ removeObject:[NSValue valueWithPointer:listener]];
+}
+
+- (void)appCheckTokenDidChangeNotification:(NSNotification*)notification {
+  NSDictionary* userInfo = notification.userInfo;
+  // Note: The Notification contains a token string, but does not contain
+  // expiration time. Could maybe call GetToken here to fetch the actual
+  // token, but there would be some risk of an infinite loop of calling
+  // GetToken and receiving TokenChanged notifications
+  NSString* app_name = (NSString*)userInfo[kFIRAppCheckAppNameNotificationKey];
+  NSString* token = (NSString*)userInfo[kFIRAppCheckTokenNotificationKey];
+  firebase::app_check::AppCheckToken cpp_token;
+  cpp_token.token = firebase::util::NSStringToString(token);
+
+  for (NSValue* listener_value in listeners_) {
+    firebase::app_check::AppCheckListener* listener =
+        static_cast<firebase::app_check::AppCheckListener*>(listener_value.pointerValue);
+    listener->OnAppCheckTokenChanged(cpp_token);
+  }
+}
+
+@end
+
 namespace firebase {
 namespace app_check {
 namespace internal {
 
 AppCheckInternal::AppCheckInternal(App* app) : app_(app) {
   future_manager().AllocFutureApi(this, kAppCheckFnCount);
   impl_ = MakeUnique<FIRAppCheckPointer>([FIRAppCheck appCheck]);
+  notification_center_wrapper_ = MakeUnique<AppCheckNotificationCenterWrapperPointer>(
+      [[AppCheckNotificationCenterWrapper alloc] init]);
```

Comment:
You likely need to unsubscribe in the destructor.  That might be why you are getting multiple responses back in the tests.

Reply:
Done. I also updated the test's TearDown to delete the AppCheck instance which will internally unsubscribe the notification center observer.
After unsubscribing the notification observer, I am seeing the correct number of notifications in integration tests.